File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        int runningRight[n];
        int runningLeft[n];
        int pR {n-1};
        for(int i = 0;i<n;i++){
            if(i==0){
                runningRight[0] = 1;
            }else{
                runningRight[i] = runningRight[i-1] * nums[i-1];
            }
            if(pR == n-1){
                runningLeft[pR] = 1;
            }else{
                runningLeft[pR] = runningLeft[pR+1] * nums[pR+1];
            }
            pR--;
        }
        for(int i = 0;i<n;i++){
            nums[i] = runningRight[i] * runningLeft[i];
        }
        return nums;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];
        int[] right = new int[n];
        ans[0] = 1;
        right[n-1] = 1;
        for(int i = 1;i<n;i++){
            ans[i] = ans[i-1] * nums[i-1];
            
        }
        for(int i = n-2;i>=0;i--){
            right[i] = right[i+1] * nums[i+1];
            ans[i+1] = right[i+1]*ans[i+1];
        }
        ans[0] = ans[0] * right[0];
        return ans;
    }
}